searx/engines/__init__.py

```diff
@@ -1,169 +1,225 @@
+    "categories": ["general"],
+    "supported_languages": [],
+    "language_aliases": {},
+    "paging": False,
+    "safesearch": False,
+    "time_range_support": False,
+    "enable_http": False,
+    "display_error_messages": True,
+    "tokens": [],
+}
+"""Defaults for the namespace of an engine module, see :py:func:`load_engine`"""
 
 categories = {'general': []}
+engines = {}
+engine_shortcuts = {}
+"""Simple map of registered *shortcuts* to name of the engine (or ``None``).
 
-babel_langs = [lang_parts[0] + '-' + lang_parts[-1] if len(lang_parts) > 1 else lang_parts[0]
-               for lang_parts in (lang_code.split('_') for lang_code in locale_identifiers())]
+::
 
-engine_shortcuts = {}
-engine_default_args = {'paging': False,
-                       'categories': ['general'],
-                       'supported_languages': [],
-                       'safesearch': False,
-                       'timeout': settings['outgoing']['request_timeout'],
-                       'shortcut': '-',
-                       'disabled': False,
-                       'enable_http': False,
-                       'time_range_support': False,
-                       'engine_type': 'online',
-                       'display_error_messages': True,
-                       'tokens': []}
+    engine_shortcuts[engine.shortcut] = engine.name
 
+"""
 
 def load_engine(engine_data):
+    """Load engine from ``engine_data``.
+
+    :param dict engine_data:  Attributes from YAML ``settings:engines/<engine>``
+    :return: initialized namespace of the ``<engine>``.
+
+    1. create a namespace and load module of the ``<engine>``
+    2. update namespace with the defaults from :py:obj:`ENGINE_DEFAULT_ARGS`
+    3. update namespace with values from ``engine_data``
+
+    If engine *is active*, return namespace of the engine, otherwise return
+    ``None``.
+
+    This function also returns ``None`` if initialization of the namespace fails
+    for one of the following reasons:
+
+    - engine name contains underscore
+    - engine name is not lowercase
+    - required attribute is not set :py:func:`is_missing_required_attributes`
+
+    """
+
     engine_name = engine_data['name']
     if '_' in engine_name:
         logger.error('Engine name contains underscore: "{}"'.format(engine_name))
-        sys.exit(1)
+        return None
 
     if engine_name.lower() != engine_name:
         logger.warn('Engine name is not lowercase: "{}", converting to lowercase'.format(engine_name))
         engine_name = engine_name.lower()
         engine_data['name'] = engine_name
 
+    # load_module
     engine_module = engine_data['engine']
-
     try:
-        engine = load_module(engine_module + '.py', engine_dir)
+        engine = load_module(engine_module + '.py', ENGINE_DIR)
     except (SyntaxError, KeyboardInterrupt, SystemExit, SystemError, ImportError, RuntimeError):
         logger.exception('Fatal exception in engine "{}"'.format(engine_module))
         sys.exit(1)
```

Comment:
Still one `sys.exit(1)` here otherwise it is possible to have a syntax error in the engine code.